relay-system/src/service.rs

```diff
@@ -44,45 +116,313 @@ impl fmt::Display for SendError {
 
 impl std::error::Error for SendError {}
 
-/// The address for a [`Service`].
+/// Response behavior of an [`Interface`] message.
 ///
-/// The address of a [`Service`] allows you to [send](Addr::send) messages to the service as
-/// long as the service is running.  It can be freely cloned.
-#[derive(Debug)]
-pub struct Addr<S: Service> {
-    /// The transmitter of the channel used to communicate with the Service
-    pub tx: mpsc::UnboundedSender<S::Messages>,
+/// The action type defines the response behavior of a service, such as asynchronous responses
+/// or fire-and-forget without responding. [`FromMessage`] implementations declare this behavior
+/// on the interface.
+///
+/// See [`FromMessage`] for more information on how to use this trait.
+pub trait MessageResponse {
+    /// Sends responses from the service back to the waiting recipient.
+    type Sender;
+
+    /// The type returned from [`Addr::send`].
+    ///
+    /// This type can be either synchronous and asynchronous based on the responder.
+    type Output;
+
+    /// Returns the response channel for an interface message.
+    fn channel() -> (Self::Sender, Self::Output);
 }
 
-// Manually derive clone since we do not require `S: Clone` and the Clone derive adds this
+/// The request when sending an asynchronous message to a service.
+///
+/// This is returned from [`Addr::send`] when the message responds asynchronously through
+/// [`AsyncResponse`]. It is a future that should be awaited. The message still runs to
+/// completion if this future is dropped.
+pub struct Request<T>(oneshot::Receiver<T>);
+
+impl<T> fmt::Debug for Request<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Request").finish_non_exhaustive()
+    }
+}
+
+impl<T> Future for Request<T> {
+    type Output = Result<T, SendError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> std::task::Poll<Self::Output> {
+        // Pinning is structural for the wrapped oneshot receiver, it implements `Unpin`.
+        // See: https://doc.rust-lang.org/stable/std/pin/index.html#pinning-is-structural-for-field
+        unsafe { self.map_unchecked_mut(|r| &mut r.0).poll(cx) }.map(|r| r.map_err(|_| SendError))
```

Comment:
Do you need unsafe?  This also always uses the field as structurally pinned:
```suggestion
        let this = &mut *self;
        Pin::new(&mut this.0)
            .poll(cx)
            .map(|r| r.map_err(|_| SendError))
```
Performance-wise I kind of expect the compiler to optimise this the same way as `Pin::map_unchecked_mut` as IIRC Pin is mostly a marker trait, but I must admit I never verified this.

I think thanks to modern auto-deref rules you don't even need the `this` part, but it makes it a little more explicit I guess.

Reply:
I think it's not possible to deref like this since there's no access to `&mut Self` without unsafe code. The docs linked in the comment elaborate on this, I think. Let me try, though.